src/error_handling.c

```diff
@@ -14,313 +14,150 @@
-#if RCUTILS_REPORT_ERROR_HANDLING_ERRORS
-    // rcutils_allocate failed, but fwrite might work?
-    RCUTILS_SAFE_FWRITE_TO_STDERR(
-      "[rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__)
-      "] failed to allocate memory for the error string\n");
-#endif
-    return;
-  }
-  snprintf(
-    __rcutils_error_string, bytes_it_would_have_written + 1,
-    __error_format_string,
-    __rcutils_error_state->message,
-    __rcutils_error_state->file,
-    __rcutils_error_state->line_number);
-  // The Windows version of snprintf does not null terminate automatically in all cases.
-  __rcutils_error_string[bytes_it_would_have_written] = '\0';
+  return str1 == str2 || 0 == strncmp(str1, str2, count);
 }
 
-const char *
-rcutils_get_error_string(void)
+void
+rcutils_set_error_state(
+  const char * error_string,
+  const char * file,
+  size_t line_number)
 {
-#ifdef RCUTILS_THREAD_LOCAL_PTHREAD
-  char * __rcutils_error_string = (char *)pthread_getspecific(__rcutils_error_string_key);
-#endif
-  if (NULL == __rcutils_error_string) {
-    format_error_string();
-  }
-  return __rcutils_error_string;
-}
+  rcutils_error_state_t error_state;
 
-bool
-__rcutils_error_is_set(rcutils_error_state_t * error_state)
-{
-  return error_state != NULL;
+  // The const is casted away to set the 'message' field of the error state (const for users only).
+  __rcutils_copy_string((char *)error_state.message, sizeof(error_state.message), error_string);
+  __rcutils_copy_string((char *)error_state.file, sizeof(error_state.file), file);
+  error_state.line_number = line_number;
+#if RCUTILS_REPORT_ERROR_HANDLING_ERRORS
+  // Only warn of overwritting if the new error is different from the old ones.
+  size_t characters_to_compare = strnlen(error_string, RCUTILS_ERROR_MESSAGE_MAX_LENGTH);
+  // assumption is that message length is <= max error string length
+  static_assert(
+    sizeof(gtls_rcutils_error_state.message) <= sizeof(gtls_rcutils_error_string.str),
+    "expected error state's max message length to be less than or equal to error string max");
+  if (
+    gtls_rcutils_error_is_set &&
+    __same_string(error_string, gtls_rcutils_error_string.str, characters_to_compare) &&
+    __same_string(error_string, gtls_rcutils_error_state.message, characters_to_compare))
+  {
+    fprintf(
+      stderr,
+      "\n"
+      ">>> [rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__) "] rcutils_set_error_state()\n"
+      "This error state is being overwritten:\n"
+      "\n"
+      "  '%s'\n"
+      "\n"
+      "with this new error message:\n"
+      "\n"
+      "  '%s, at %s:%zu'\n"
+      "\n"
+      "rcutils_reset_error() should be called after error handling to avoid this.\n"
+      "<<<\n",
+      gtls_rcutils_error_string.str,
+      error_string,
+      file,
+      line_number);
+  }
+#endif
+  gtls_rcutils_error_state = error_state;
+  gtls_rcutils_error_string_is_formatted = false;
+  gtls_rcutils_error_string = (const rcutils_error_string_t) {0};
```

Comment:
I don't actually know this syntax in C; what does this do to a statically-allocated array embedded in a structure?